[flang][debug] Generate DWARF debug info using fir.use_stmt.

abidh opened the pull request:
This patch uses the fir.use_stmt operations to generate correct debug metadata for use statement when `only` and `=>` are used. The debug flow is changed a bit where we process the module globals first so that we have the global variables when we start to process `fir.use_stmt`.
    
Fixes #160923.

abidh committed:
[flang] Generate DWARF debug info using fir.use_stmt.

This patch uses the fir.use_stmt operations to generate correct
debug metadata for use statement when `only` and `=>` are used.

The debug flow is changed a bit where we process the module globals
first so that we have the global variables when we start to process
fir.use_stmt.

Fixes #160923.

abidh committed:
Rename a testfile.

tblah reviewed (approved):
LGTM with some very minor nits.

It would be best if this could also be reviewed by somebody with more expertise in DWARF than I have, but in the absence of such a reviwer I think this looks good so far as I can tell.

jeanPerier reviewed (approved):
Also not a dwarf expert but the code looks good, nothing to add over Tom's comments. Thanks!

abidh committed:
Handle review comments.

1. Move 2 helps outisde anonymoud namespace and add static.
2. Refactor some duplicate code into a helper.

tblah reviewed (approved):
Thanks for the update. LGTM
